File: LectureCodes/Interpolation/num_diff_fd/Eigen/numericaldifferentiation.cpp
# include <vector>
# include <cmath>
# include <figure/figure.hpp>

# define SCALAR long double

template <typename Function>
void diff(const double& x, Function& f, Function& df) {
  std::vector<SCALAR> error, h;
  // build vector of difference quotient width
  for (int i = -1; i >= -61; i -= 5) {
    h.push_back( std::pow(2, i) );
  }
  for (int j = 0; j < h.size(); ++j) {
    const double df_approx = (f(x + h[j]) - f(x))/h[j];
    error.push_back( std::abs(df_approx - df(x)) );
  }

  mgl::Figure fig;
  fig.title("One sided-diff.quotient");
  fig.xlabel("h");
  fig.ylabel("error");
  fig.setlog(true, true);
  fig.plot(h, error, "+-");
  fig.save("expnumdiff.eps");
}

int main() {
  auto f = [](double x){ return std::exp(x); };
  auto df = f;
  diff(1.1, f, df);
  return 0;
}

File: LectureCodes/MatVec/Dense/expnumdiff/Matlab/numericaldifferentiation.m
function experr = numericaldifferentiation()
% Numerical differentiation of exponential function with extended precision arithmetic
% Uses the Advanpix extended precision library for MATLAB, wwww.advanpix.com

experr = []; l = 1;
for ndigits=[5 10 15 20 25 30 35]
  mp.Digits(ndigits), % Set number of digits
  x = mp('1.1'); % Evaluation point in extended precision 
  h = mp(2).^[-1:-5:-61]; % width of difference quotient in extended precision
  experr = [experr; abs(((exp(x+h)-exp(x))./h) - exp(x))]; % compute (absolute) error
  leg{l} = sprintf('%2.0d digits',ndigits);
  l = l+1;
end
  
% Graphical output of error
figure('name','numdiff');
loglog(h,experr,'+-');
title('One-sided difference quotient approximation of derivative of e^x');
xlabel('h','fontsize',14);
ylabel('error','fontsize',14);
legend(leg,'Location','best');
print -depsc2 'expnumdiffmultiprecision.eps';
